kythe/docs/rfc/2967.md

```diff
@@ -0,0 +1,64 @@
+## Build Configuration in the Kythe Graph
+
+[TOC]
+
+## Objective {#objective}
+
+Incorporate build configuration information into the Kythe graph. Somewhat relatedly, provide canonical guidelines on how to identify build-time generated artifacts.
+
+
+## Background {#background}
+
+Currently, the Kythe graph largely assumes that it is built for a single configuration, at a consistent revision for a particular corpus.  Historically, this has worked well.  But as time marches onward and the diversity of platforms built from a single corpus has grown it has left a widening gap in the information available to users with only partial cross references for some files (due to conditional compilation) or other files entirely omitted due to build-time selection.  In order to work around this lack, some third-party corpora (most notably Chromium) have begun to use the "root" VName field to encode this information.  Similarly, the extant "prodkernel" corpus has need of this, but currently makes do without. We reasonably anticipate that other soon-to-be-corpora such as Android and Tensorflow would encounter similar limitations.
+
+
+## Overview {#overview}
+
+Rather than introduce any new VName fields, nodes or edges build configuration can be adequately represented via a signature-relevant fact on anchor nodes and some ancillary clarifications to the guidelines for root and corpus.  This is predicated on:
```

Comment:
The Background above talks about build configuration, but I don't see anything that discusses the specific problem that causes for the graph at the node level. In other words, it'd be helpful to answer the question "why is _this_ solution preferred?" in addition to "what are we planning to do?"

Chromium is using the root in that way because they have to serve all their xrefs from a UI that doesn't know how to distinguish "variants" like platform strings, talking to a service that only serves a single corpus. But that's an historical accident. The question I see here is—even if the code browser _did_ know how to do that, why does that mean we have to add data to the _graph_ for that information? 

Using C++ again (running with the Chromium example), each of the build configurations has to be built and indexed separately anyway, since they use different flags, include-paths, etc. Assuming we fixed the problem that the UI doesn't have a knob for the platform—could this problem be as effectively solved by storing the graph for (say) the "android" variant separately from the graph for (say) the "linux" variant?

In other words: What do we get out of stuffing this information into the graph at the node level, that we could not get by having multiple graphs?  The list below mentions filtering—could that not be done by selecting from a set of graphs rather than examining each node individually? (more on this below)

(I assume there is a reason for this, but it isn't spelled out anywhere that I can see)

Reply:
The answer to your first question is the first bullet point "Non-generated source files exist independent of build configuration."  If the data is in a different graph for each build configuration every xrefs request has to be split across each build configuration.  And then stitched together either at the client or in a proxy.  It's inefficient. This also applies to graph storage generally; using distinct graphs means we have N copies of essentially the same data.

C++ is the poster child because a single file is likely to have different anchors (possible to the same underlying semantic entities) depending on the build configuration.  Keeping them in the same graph allows us to provide references to each distinct entity, regardless of platform (e.g. header w/ multiple implementations).  Splitting the graph does not allow for this.

Reply:
Figuring out all of the xrefs for a "particular" semantic entity will also be lossy as that entity will potentially have different tickets in each graph and the only key that they are the "same" being the fact that the anchors happen to overlap.

Reply:
> If the data is in a different graph for each build configuration every xrefs request has to be split across each build configuration.

Is there any expectation that xrefs will be shared _across_ build configurations? In C++, for example, that will ~never happen because of how tickets are generated.

Reply:
I think those points would be helpful to include in the motivation for the design.